src/activatable/activatable.ts

```diff
@@ -0,0 +1,130 @@
+/**
+ * Created ->- Activating ->- Activated ->- Deactivating ->- Deactivated -------------->----------- Disposed
+ *                 |  |                                         |  |                                  |  |
+ *                 |  `------------------<----------------------'  `--->--- FailedToDeactivate --->---'  |
+ *                 `--------------------->--------- FailedToActivate ------------------>-----------------'
+ */
+
+enum LifeCycleState {
+  Created = 1,
+  Activating,
+  Activated,
+  Deactivating,
+  Deactivated,
+  Disposed,
+  FailedToActivate,
+  FailedToDeactivate,
+}
+
+const LifeCycleDescription = {
+  [LifeCycleState.Created]: 'has never activated before',
+  [LifeCycleState.Activating]: 'is activating',
+  [LifeCycleState.Activated]: 'is activated',
+  [LifeCycleState.Deactivating]: 'is deactivating',
+  [LifeCycleState.Deactivated]: 'is deactivated',
+  [LifeCycleState.Disposed]: 'is disposed',
+  [LifeCycleState.FailedToActivate]: 'errored while activating',
+  [LifeCycleState.FailedToDeactivate]: 'errored while deactivating',
+};
+
+export interface IActivatable {
+  isActivating: boolean;
+  isActivated: boolean;
+  isDeactivating: boolean;
+  isDeactivated: boolean;
+  isErrored: boolean;
+  activate: () => Promise<void>;
+  deactivate: () => Promise<void>;
+  disposeActivateable: () => void;
+}
+
+export abstract class Activatable implements IActivatable {
+  private _state: LifeCycleState = LifeCycleState.Created;
```

Comment:
Any reason for not making this a public property?

Reply:
because then you could violate the state machine edges. I could add a read-only public getter for it?

Reply:
> I could add a read-only public getter for it?

You don't need a getter. You can just have a decorator that would add a non-writable property.

Reply:
... but I know how to write a getter lol.

Reply:
We could use https://www.npmjs.com/package/core-decorators#readonly

Reply:
you are just trolling me now.

Reply:
lets not introduce a new library for a very simple pattern that we already use extensively because of MobX `@computed` getters.

Reply:
I'm not :thinking:.
I'd really rather expose a single property called `state`, than these `isSomething` properties that hamper DX in the end code.

Reply:
Anyhoo, we can have
```ts
private _state:  LifeCycleState = LifeCycleState.Created;

public get state() {
  return  this._state;
}
```

Reply:
> I could add a read-only public getter for it?

20 comments later...

> Anyhoo, we can have
> 
> ```ts
> private _state:  LifeCycleState = LifeCycleState.Created;
> 
> public get state() {
>   return  this._state;
> }
> ```

Reply:
:rofl: :rofl: